kernel/mellanox_ethernet/src/command_queue.rs

```diff
@@ -0,0 +1,740 @@
+    InitHca                 = 0x102,
+    TeardownHca             = 0x103,
+    EnableHca               = 0x104,
+    DisableHca              = 0x105,
+    QueryPages              = 0x107,
+    ManagePages             = 0x108,
+    QueryIssi               = 0x10A,
+    SetIssi                 = 0x10B,
+    QuerySpecialContexts    = 0x203,
+    CreateEq                = 0x301,
+    CreateCq                = 0x400,
+    QueryVportState         = 0x751,
+    QueryNicVportContext    = 0x754,
+    AllocUar                = 0x802,
+    AllocPd                 = 0x800,
+    AllocTransportDomain    = 0x816,
+    CreateTis               = 0x912,
+    CreateSq                = 0x904,
+    ModifySq                = 0x905,
+    CreateRq                = 0x908,
+    ModifyRq                = 0x909,
+    Unknown
+}
+
+#[derive(Debug)]
+/// Command status written by HW in status field of the output data in the command entry.
+pub enum CommandReturnStatus {
+    OK                  = 0x00,
+    InternalError       = 0x01,
+    BadOp               = 0x02,
+    BadParam            = 0x03,
+    BadSysState         = 0x04,
+    BadResource         = 0x05,
+    ResourceBusy        = 0x06,
+    ExceedLim           = 0x08,
+    BadResState         = 0x09,
+    BadIndex            = 0x0A,
+    NoResources         = 0x0F,
+    BadInputLen         = 0x50,
+    BadOutputLen        = 0x51,
+    BadResourceState    = 0x10,
+    BadPkt              = 0x30,
+    BadSize             = 0x40,
+    Unknown
+}
+
+/// Possible values of the opcode modifer when the opcode is ManagePages
+pub enum ManagePagesOpmod {
+    AllocationFail      = 0,
+    AllocationSuccess   = 1,
+    HcaReturnPages      = 2
+}
+
+/// Possible values of the opcode modifer when the opcode is QueryPages
+pub enum QueryPagesOpmod {
+    BootPages       = 1,
+    InitPages       = 2,
+    RegularPages    = 3
+}
+
+#[derive(PartialEq)]
+/// Mailboxes can be used for both input data passed to HW, and output data passed from HW to SW.
+enum MailboxType {
+    Input,
+    Output
+}
+
+/// A buffer of fixed-size entries that is used to pass commands to the HCA.
+/// It resides in a physically contiguous 4 KiB memory chunk.
+/// (Section 8.24.1: HCA Command Queue)
+#[repr(C)]
+pub struct CommandQueue {
+    /// Physically-contiguous command queue entries
+    entries: BoxRefMut<MappedPages, [CommandQueueEntry]>,
+    /// Per-entry boolean flags to keep track of which entries are in use
+    available_entries: Vec<bool>,
+    /// A random number that needs to be different for every command, and the same for all mailboxes that are part of a command.
+    token: u8,
+    /// Pages used for input mailboxes 
+    mailbox_buffers_input: Vec<Vec<(MappedPages, PhysicalAddress)>>, 
+    /// Pages used for output mailboxes 
+    mailbox_buffers_output: Vec<Vec<(MappedPages, PhysicalAddress)>> 
+}
+
+impl CommandQueue {
+
+    /// Create a command queue object.
+    ///
+    /// # Arguments
```

Comment:
Moving forward, let's use `##` instead of `#` for section headers in doc comments. It's a lot easier to read and seems to be what the rust stdlib does.